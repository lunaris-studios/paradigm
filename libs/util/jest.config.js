module.exports = {
	name: "util",
	preset: "../../jest.config.js",
	transform: {
		"^.+\\.[tj]sx?$": [
			"babel-jest",
			{ cwd: __dirname, configFile: "./babel-jest.config.json" },
		],
	},
	moduleNameMapper: {
		"^$util/(.*)$": "<rootDir>/src/$1",
		"^lodash-es$": "lodash",
	},
	moduleFileExtensions: ["ts", "tsx", "js", "jsx", "html"],
	coverageDirectory: "../../coverage/libs/util",
};
