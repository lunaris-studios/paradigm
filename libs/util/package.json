{
	"name": "@paradigmjs/util",
	"version": "0.0.17",
	"description": "Abstract React Components",
	"repository": {
		"type": "git",
		"url": "git@github.com:lunaris-studios/paradigm.git",
		"directory": "packages/util"
	},
	"license": "MIT",
	"author": "Sam Craig <sam@lunaris.io>",
	"peerDependencies": {
		"react": "^16.13.1",
		"react-dom": "^16.13.1"
	},
	"publishConfig": {
		"access": "public",
		"directory": "../../dist/libs/util"
	}
}
