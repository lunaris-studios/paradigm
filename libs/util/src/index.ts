export * from "./ts";
export * from "./util";

export * from "utility-types";
export * from "lodash-es";
