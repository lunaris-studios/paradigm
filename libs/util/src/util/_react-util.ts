import * as React from "react";

import { isFunction } from "lodash-es";

/**
 * Returns true if `node` is null/undefined, false, empty string, or an array
 * composed of those. If `node` is an array, only one level of the array is
 * checked, for performance reasons.
 */
export function isReactNodeEmpty(node?: React.ReactNode, skipArray = false): boolean {
	return (
		node == null ||
		node === "" ||
		node === false ||
		(!skipArray &&
			Array.isArray(node) &&
			// only recurse one level through arrays, for performance
			(node.length === 0 || node.every((n) => isReactNodeEmpty(n, true))))
	);
}

/**
 * Converts a React node to an element: non-empty string or number or
 * `React.Fragment` (React 16.3+) is wrapped in given tag name; empty strings
 * and booleans are discarded.
 */
export function ensureElement(
	child: React.ReactNode | undefined,
	tagName: keyof JSX.IntrinsicElements = "span",
) {
	if (child == null || typeof child === "boolean") {
		return undefined;
	} else if (typeof child === "string") {
		// cull whitespace strings
		return child.trim().length > 0 ? React.createElement(tagName, {}, child) : undefined;
	} else if (
		typeof child === "number" ||
		typeof (child as any).type === "symbol" ||
		Array.isArray(child)
	) {
		// React.Fragment has a symbol type, ReactNodeArray extends from Array
		return React.createElement(tagName, {}, child);
	} else if (isReactElement(child)) {
		return child;
	} else {
		// child is inferred as {}
		return undefined;
	}
}

/**
 * Returns true if `node` is of type <ReactElement>. If `node` is not,
 * returns false.
 */
export function isReactElement<T = any>(
	child: React.ReactNode,
): child is React.ReactElement<T> {
	return (
		typeof child === "object" &&
		typeof (child as any).type !== "undefined" &&
		typeof (child as any).props !== "undefined"
	);
}

/**
 * Represents anything that has a `name` property such as Functions.
 */
export interface INamed {
	name?: string;
}

/**
 * Returns the display name of a React Component if available.

 * @param ComponentClass React Component element in question
 */
export function getDisplayName(ComponentClass: React.ComponentType | INamed) {
	return (
		(ComponentClass as React.ComponentType).displayName ||
		(ComponentClass as INamed).name ||
		"Unknown"
	);
}

/**
 * Returns true if the given JSX element matches the given component type.
 *
 * NOTE: This function only checks equality of `displayName` for performance and
 * to tolerate multiple minor versions of a component being included in one
 * application bundle.
 * @param element JSX element in question
 * @param ComponentType desired component type of element
 */
export function isElementOfType<P = {}>(
	element: any,
	ComponentType: React.ComponentType<P>,
): element is React.ReactElement<P> {
	return (
		element != null &&
		element.type != null &&
		element.type.displayName != null &&
		element.type.displayName === ComponentType.displayName
	);
}

/**
 * Given the version number MAJOR.MINOR.PATCH of the installed React library,
 * returns the MAJOR version integer.
 */
export function getReactMajorVersion(): number {
	const majorVersionString = React.version.substr(0, React.version.indexOf("."));
	const majorVersionNumber = parseInt(majorVersionString);
	return majorVersionNumber;
}

/**
 * Given the version number MAJOR.MINOR.PATCH of the installed React library,
 * returns true if the MAJOR version integer is greater than or equal
 * to version `16`.
 */
export function isReact16Plus(): boolean {
	const majorVerison = getReactMajorVersion();
	return Boolean(majorVerison >= 16);
}

/**
 * Returns type `T` if extension of `render` is valid.
 */
type HasRenderProp<T> = T extends { render: (props: any) => React.ReactNode } ? T : never;

/**
 * Returns true if component props contains `render`.
 */
export function hasRender<T extends {}>(props: T): props is HasRenderProp<T> {
	return "render" in props && isFunction((props as HasRenderProp<T>).render);
}

/**
 * Returns type `T` if extension of `children` is valid.
 */
type HasChildrenProp<T> = T extends { children: (props: any) => React.ReactNode }
	? T
	: never;

/**
 * Returns true if component props contains `children`.
 */
export function hasChildren<T extends {}>(props: T): props is HasChildrenProp<T> {
	return "children" in props && isFunction((props as HasChildrenProp<T>).children);
}

/** A collection of curated prop keys used across our Components which are not valid HTMLElement props. */
const INVALID_PROPS = [
	"active",
	"alignText",
	"containerRef",
	"current",
	"elementRef",
	"fill",
	"icon",
	"inputRef",
	"intent",
	"inline",
	"large",
	"loading",
	"leftElement",
	"leftIcon",
	"minimal",
	"onRemove", // ITagProps, ITagInputProps
	"outlined", // IButtonProps
	"panel", // ITabProps
	"panelClassName", // ITabProps
	"popoverProps",
	"rightElement",
	"rightIcon",
	"round",
	"small",
	"text",
];

/**
 * Typically applied to HTMLElements to filter out blacklisted props. When applied to a Component,
 * can filter props from being passed down to the children. Can also filter by a combined list of
 * supplied prop keys and the blacklist (only appropriate for HTMLElements).
 * @param props The original props object to filter down.
 * @param {InvalidProps[]} invalidProps If supplied, overwrites the default blacklist.
 * @param {boolean} shouldMerge If true, will merge supplied invalidProps and blacklist together.
 */
export function removeNonHTMLProps(
	props: { [key: string]: any },
	invalidProps = INVALID_PROPS,
	shouldMerge = false,
): { [key: string]: any } {
	if (shouldMerge) {
		invalidProps = invalidProps.concat(INVALID_PROPS);
	}

	return invalidProps.reduce(
		(prev, curr) => {
			if (prev.hasOwnProperty(curr)) {
				delete prev[curr];
			}

			return prev;
		},
		{ ...props },
	);
}
