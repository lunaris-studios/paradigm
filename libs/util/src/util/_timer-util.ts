/**
 * requestInterval() & requestTimeout() by Joe Lambert (@joelambert)
 * https://gist.github.com/joelambert/1002116
 */

import { default as raf } from "raf";
import * as Common from "$util/common";

export const foo = Common.DISPLAYNAME_PREFIX;

export interface IRequestHandle {
	id: number;
}

export type RequestFn = () => void;

/**
 * Behaves the same as setInterval except uses requestAnimationFrame()
 * where possible for better performance.
 * @param {Function} fn The callback function
 * @param {number} interval The interval in milliseconds
 */
export function requestInterval(fn: RequestFn, interval: number) {
	const handle = {} as IRequestHandle;
	let start = new Date().getTime();

	function loop() {
		const current = new Date().getTime();
		const delta = current - start;

		if (delta >= interval) {
			fn();
			start = new Date().getTime();
		}

		handle.id = raf(loop);
	}

	handle.id = raf(loop);
	return handle;
}

/**
 * Behaves the same as clearInterval except uses cancelRequestAnimationFrame()
 * where possible for better performance.
 * @param {number} id ID of the requestInterval() handle
 */
export function clearRequestInterval(id: number) {
	raf.cancel(id);
}

/**
 * Behaves the same as setTimeout except uses requestAnimationFrame()
 * where possible for better performance.
 * @param {Function} fn The callback function
 * @param {number} timeout The delay in milliseconds
 */

export function requestTimeout(fn: RequestFn, timeout: number) {
	const handle = {} as IRequestHandle;
	let start = new Date().getTime();

	function loop() {
		const current = new Date().getTime();
		const delta = current - start;

		if (delta >= timeout) {
			fn();
		} else {
			handle.id = raf(loop);
		}
	}

	handle.id = raf(loop);
	return handle;
}

/**
 * Behaves the same as clearTimeout except uses cancelRequestAnimationFrame()
 * where possible for better performance.
 * @param {number} id ID of the requestTimeout() handle
 */
export function clearRequestTimeout(id: number) {
	raf.cancel(id);
}

/**
 * Instantiates an empty promise to block a async execution stack, resolving
 * after the provided `delay` in milliseconds.
 * @param {number} delay Time in milliseconds.
 */
export function wait(delay: number) {
	return new Promise((r) => setTimeout(r, delay));
}
