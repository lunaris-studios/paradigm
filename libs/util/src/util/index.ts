export * from "./_function-util";
export * from "./_js-util";
export * from "./_react-util";
export * from "./_timer-util";
