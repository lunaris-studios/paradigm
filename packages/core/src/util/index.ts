export * from "./_can-use-dom";
export * from "./_compose";
export * from "./_function-utils";
export * from "./_is-key-board-click";
export * from "./_is-node-env";
export * from "./_js-utils";
export * from "./_pipe";
export * from "./_react-utils";
export * from "./_remove-non-html-props";
export * from "./_test-util";

import clamp from "lodash/clamp";

export { clamp };
export { default as isNull } from "lodash/isNull";
export { default as camelCase } from "lodash/camelCase";
