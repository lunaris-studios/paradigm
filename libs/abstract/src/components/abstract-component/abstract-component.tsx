import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";

import * as Util from "$abstract/util";

/**
 * An abstract component that Paradigm components can extend
 * in order to add some common functionality like runtime props validation.
 */
export abstract class AbstractComponent<
	TProps,
	TState = Record<string, unknown>,
	TSnapshot = Record<string, unknown>
> extends React.Component<TProps, TState, TSnapshot> {
	// this should be static, not an instance method
	public getDerivedStateFromProps: never;

	// Not bothering to remove entries when their timeouts finish because clearing invalid ID is a no-op
	private timeoutIds: number[] = [];

	private requestTimeoutIds: number[] = [];

	constructor(props: TProps) {
		super(props);
		if (!Util.isNodeEnv(Protocol.Stage.PRODUCTION) && this.validateProps) {
			this.validateProps(this.props);
		}
	}

	public componentDidUpdate(
		_prevProps: TProps,
		_prevState: TState,
		_snapshot?: TSnapshot,
	): void {
		if (!Util.isNodeEnv(Protocol.Stage.PRODUCTION) && this.validateProps) {
			this.validateProps(this.props);
		}
	}

	public componentWillUnmount(): void {
		this.clearTimeouts();
		this.clearRequestTimeouts();
	}

	//

	/**
	 * Set a timeout and remember its ID.
	 * All stored timeouts will be cleared when component unmounts.
	 * @returns a "cancel" function that will clear timeout when invoked.
	 */
	public setTimeout(callback: () => void, timeout = 0): Util.RequestFn {
		const handle = window.setTimeout(callback, timeout);
		this.timeoutIds.push(handle);
		return (): void => window.clearTimeout(handle);
	}

	/**
	 * Clear all known timeouts.
	 */
	public clearTimeouts(): void {
		if (this.timeoutIds.length > 0) {
			this.requestTimeoutIds.forEach((id) => {
				Util.clearRequestTimeout(id);
			});
			this.timeoutIds = [];
		}
	}

	/**
	 * Set a timeout driven by raf() and remember
	 * its ID. All stored timeouts will be cleared when component unmounts.
	 */
	public setRequestTimeout(callback: Util.RequestFn, timeout = 0): Util.RequestFn {
		const handle = Util.requestTimeout(callback, timeout);
		this.requestTimeoutIds.push(handle.id);
		return (): void => Util.clearRequestTimeout(handle.id);
	}

	/**
	 * Clear all known raf() timeouts.
	 */
	public clearRequestTimeouts(): void {
		if (this.requestTimeoutIds.length > 0) {
			this.requestTimeoutIds.forEach((id) => {
				Util.clearRequestTimeout(id);
			});
			this.requestTimeoutIds = [];
		}
	}

	/**
	 * Ensures that the props specified for a component are valid.
	 * Implementations should check that props are valid and usually throw an Error if they are not.
	 * Implementations should not duplicate checks that the type system already guarantees.
	 *
	 * This method should be used instead of React's
	 * [propTypes](https://facebook.github.io/react/docs/reusable-components.html#prop-validation) feature.
	 * Like propTypes, these runtime checks run only in development mode.
	 */
	protected validateProps?(_props: TProps): void;
}




















