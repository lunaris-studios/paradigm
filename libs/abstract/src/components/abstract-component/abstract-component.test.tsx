import * as React from "react";
import * as Testing from "@testing-library/react";

import * as Util from "$abstract/util";

import * as Component from "./abstract-component";

const TEST_TIMEOUT = 10;

describe("AbstractComponent", () => {
	interface IAbstractComponentWithTimeoutProps {}

	class AbstractComponent extends Component.AbstractComponent<
		IAbstractComponentWithTimeoutProps
	> {
		public render(): JSX.Element {
			return <div />;
		}
	}

	it("renders correctly", () => {
		const { baseElement } = Testing.render(<AbstractComponent />);
		expect(baseElement).toMatchSnapshot();
	});

	interface IAbstractComponentWithTimeoutProps {}
	interface IAbstractComponentWithTimeoutState {
		flag: number;
	}

	class AbstractComponentWithTimeout extends Component.AbstractComponent<
		IAbstractComponentWithTimeoutProps,
		IAbstractComponentWithTimeoutState
	> {
		public state: IAbstractComponentWithTimeoutState = { flag: 0 };

		public timeout = this.setTimeout(
			() =>
				this.setState({
					flag: 1,
				}),
			TEST_TIMEOUT,
		);

		public render(): JSX.Element {
			const { flag } = this.state;
			return (
				<React.Fragment>
					<span data-testid="flag">{flag}</span>
					<span data-testid="timeout">{flag}</span>
				</React.Fragment>
			);
		}
	}

	it("creates timeout", async () => {
		const { getByTestId } = Testing.render(<AbstractComponentWithTimeout />);

		expect(getByTestId("flag")).toHaveTextContent("0");

		await Testing.waitFor((): void => {
			expect(getByTestId("flag")).toHaveTextContent("1");
		});
	});

	it("active timeout can be cancelled", async () => {
		const { getByTestId } = Testing.render(<AbstractComponentWithRAFTimeout />);

		expect(getByTestId("flag")).toHaveTextContent("0");

		Testing.fireEvent.click(getByTestId("timeout"));

		Util.wait(TEST_TIMEOUT);

		expect(getByTestId("flag")).not.toHaveTextContent("1");
	});

	interface IAbstractComponentWithRAFTimeoutProps {}
	interface IAbstractComponentWithRAFTimeoutState {
		flag: number;
	}

	class AbstractComponentWithRAFTimeout extends Component.AbstractComponent<
		IAbstractComponentWithRAFTimeoutProps,
		IAbstractComponentWithRAFTimeoutState
	> {
		constructor(props: IAbstractComponentWithRAFTimeoutProps) {
			super(props);

			this.timeout = this.timeout.bind(this);
		}

		public state: IAbstractComponentWithRAFTimeoutState = { flag: 0 };

		public timeout = this.setRequestTimeout(
			() =>
				this.setState({
					flag: 1,
				}),
			TEST_TIMEOUT,
		);

		public render(): JSX.Element {
			const { flag } = this.state;
			return (
				<React.Fragment>
					<span data-testid="flag">{flag}</span>
					<button
						aria-label="timeout"
						data-testid="timeout"
						onClick={this.timeout}
						onKeyDown={this.timeout}
						type="button"
					/>
				</React.Fragment>
			);
		}
	}

	it("creates raf() timeout", async () => {
		const { getByTestId } = Testing.render(<AbstractComponentWithRAFTimeout />);

		expect(getByTestId("flag")).toHaveTextContent("0");

		await Testing.waitFor(() => {
			expect(getByTestId("flag")).toHaveTextContent("1");
		});
	});

	it("active raf() timeout can be cancelled", async () => {
		const { getByTestId } = Testing.render(<AbstractComponentWithRAFTimeout />);

		expect(getByTestId("flag")).toHaveTextContent("0");

		Testing.fireEvent.click(getByTestId("timeout"));

		Util.wait(TEST_TIMEOUT);

		expect(getByTestId("flag")).not.toHaveTextContent("1");
	});
});
