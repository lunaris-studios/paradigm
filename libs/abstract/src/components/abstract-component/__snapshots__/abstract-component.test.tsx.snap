// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractComponent renders correctly 1`] = `
<body>
  <div>
    <div />
  </div>
</body>
`;
